base/rational.jl

```diff
@@ -129,12 +129,220 @@ Bool(x::Rational) = x==0 ? false : x==1 ? true :
+    dexp = tz_num - tz_den
+    c = rational_to_float_components_impl(
+        num >> tz_num, den >> tz_den, bit_width, T, romo, h(exp_lower_bound, dexp))
+    FloatComponentsResult(
+        c.mantissa, c.exponent + dexp, inexact = c.inexact, underflowed = c.underflowed)
+end
+
+# Assuming the wanted rounding mode is round to nearest with ties to
+# even.
+#
+# `precision` must be positive.
+function rational_to_float_impl(to_float::C,
+                                ::Type{T},
+                                x::Rational,
+                                precision::Integer) where {C<:Union{Type,Function},T<:Integer}
+    s = Int8(sign(numerator(x)))
+    a = abs(x)
+
+    num = numerator(a)
+    den = denominator(a)
+
+    # Handle special cases
+    num_is_zero = iszero(num)
+    den_is_zero = iszero(den)
+    if den_is_zero
+        num_is_zero && return to_float(NaN)  # 0/0
+        return to_float(s * Inf)  # n/0 = sign(n) * Inf
+    end
+    num_is_zero && return to_float(false)  # 0/n = 0
+
+    F = typeof(to_float(false))
+    c = rational_to_float_components(
+        num, den, precision, T, RoundNearest, min_normal_number_exponent(F))
+    bw = bit_width(c.mantissa)
+
+    # TODO: `ldexp` could be replaced with a mere bit of bit twiddling
+    # in the case of `Float16`, `Float32`, `Float64`
+    ldexp(s * to_float(c.mantissa), c.exponent - bw + true)::F
 end
 
+function rational_to_float_promote_type(::Type{F},
+                                        ::Type{S}) where {F<:AbstractFloat,S<:Integer}
+    BigInt
+end
+
+function rational_to_float_promote_type(::Type{F},
+                                        ::Type{S}) where {F<:AbstractFloat,S<:Unsigned}
+    rational_to_float_promote_type(F, widen(signed(S)))
+end
+
+# As an optimization, use a narrower type when possible.
+rational_to_float_promote_type(::Type{Float16},                          ::Type{<:Union{Int8,Int16}}) = Int32
+rational_to_float_promote_type(::Type{Float32},                          ::Type{<:Union{Int8,Int16}}) = Int64
+rational_to_float_promote_type(::Type{Float64},                          ::Type{<:Union{Int8,Int16}}) = Int128
+rational_to_float_promote_type(::Type{<:Union{Float16,Float32}},         ::Type{Int32})               = Int64
+rational_to_float_promote_type(::Type{Float64},                          ::Type{Int32})               = Int128
+rational_to_float_promote_type(::Type{<:Union{Float16,Float32,Float64}}, ::Type{Int64})               = Int128
+
+function rational_bool_to_float(to_float::C,
+                                x::Rational{Bool}) where {C<:Union{Type,Function}}
+    n = numerator(x)
+    if iszero(denominator(x))
+        if iszero(n)
+            to_float(NaN)  # 0/0
+        else
+            to_float(Inf)  # 1/0
+        end
+    else
+        # n/1 = n
+        to_float(n)
+    end
+end
+
+function rational_to_float(::Type{F}, x::Rational{Bool}) where {F<:AbstractFloat}
+    rational_bool_to_float(F, x)::F
+end
```

Comment:
No need for an additional indirection, please inline this.

Reply:
Not necessarily - the use in `base/mpfr.jl` can just call `BigFloat(numerator(x) / denominator(x))` instead, since that will always be accurate.

Come to think of it, can't we do that here as well, saving the indirection of that `to_float` wrapper entirely?